File: semILaby/laby6/zadanie3/perm.py
def findIndex(tab, x, a, b):
    i = a 
    ans = tab[a]
    ind = i 
    while i <= b:
        if tab[i] > x:
            if tab[i] < ans:
                ans = tab[i]
                ind = i 
        i++
    return ind 

'''
def factorial(n):
    ans = 1
    for i in (2 .. n + 1):
        ans *= i
    return ans
'''

def swap(tab, a, b):
    c = tab[a]
    tab[a] = tab[b]
    tab[b] = c 

def rev(tab, a, b):
    while a < b:
        swap(tab, a, b)
        a++
        b--

def nextPermutation(tab, n):
    l = n - 1
    while tab[l] < tab[l-1] and l > 0:
        #print("-1")
        l--
    if l == 0:
        return False

    i = findIndex(tab, tab[l - 1], l, n - 1)
    swap(tab, l-1, i)
    rev(tab, l, n-1)
'''
   for i in (0, n):
       print(tab[i])
'''
    return True

File: laby/sem1/laby6/zadanie1/perm.c
#include <stdio.h>
#include <stdbool.h>
int FindIndex(int tab[], int x, int a, int b)
{
	int i=a;
	int ans=tab[a];
	int ind=i;
	while(i<=b)
	{
		if(tab[i]>x)
		{
			if(tab[i]<ans)
			{
				ans=tab[i];
				ind = i;
			}
		}
		i++;
	}
	return ind;
}
/*int Factorial(int n)
{
	int ans=1;
	for(int i=2;i<=n;i++)
	{
		ans*=i;
	}
	return ans;
}*/
void Swap(int tab[], int a, int b)
{
	int c;
	c = tab[a];
	tab[a] = tab[b];
	tab[b] = c;
}
void Rev(int tab[], int a, int b)
{
	while(a<b)
	{
		Swap(tab,a,b);
		a++;
		b--;
	}
}
bool NextPermutation(int tab[], int n)
{
	int l = n-1;
	while((tab[l]<tab[l-1])&&l>0)
	{
		l--;
	}
	if(l==0)
	{
		//printf("-1");
		return false;
	}
	int i = FindIndex(tab, tab[l-1], l, n-1);
	Swap(tab, l-1, i);
	Rev(tab, l, n-1);
	/*for(int i=0;i<n;i++)
	{
		printf("%d ", tab[i]);
	}*/
	return true;
}
